File: backend/src/main/scala/slaydemo/backend/replay/objects/ReplayTypes.scala
package slaydemo.backend.replay.objects

import slaydemo.backend.battle.objects.{BattleId, BattlePlacement, BattleSurvivalOutcome, DurationMillis, EpochMillis, Rating, RatingDelta, Score}
import slaydemo.backend.identity.objects.{DisplayName, PlayerHandle}

final case class ReplayId(value: String) extends AnyVal
final case class ReplayCommentId(value: String) extends AnyVal
final case class ReplayTitle private (value: String) extends AnyVal
final case class ReplayFramesJson private (value: String) extends AnyVal
final case class ReplayFrameCount private (value: Int) extends AnyVal

object ReplayTitle {
  def fromWire(value: String): ReplayTitle =
    new ReplayTitle(Option(value).getOrElse(""))
}

object ReplayFramesJson {
  val empty: ReplayFramesJson =
    new ReplayFramesJson("[]")

  def fromNormalized(value: String): ReplayFramesJson =
    new ReplayFramesJson(value)
}

object ReplayFrameCount {
  val zero: ReplayFrameCount =
    new ReplayFrameCount(0)

  def fromWire(value: Int): ReplayFrameCount =
    new ReplayFrameCount(math.max(0, value))
}

final case class ReplaySettlementRecord(
  handle: PlayerHandle,
  displayName: DisplayName,
  resultLabel: String,
  highlightLine: String,
  score: Score,
  placement: Option[BattlePlacement],
  ratingBefore: Option[Rating],
  ratingDelta: Option[RatingDelta],
  ratingAfter: Option[Rating],
  survivalOutcome: BattleSurvivalOutcome,
  currentLoadout: Option[String]
) {
  def aliveAtEnd: Boolean =
    BattleSurvivalOutcome.aliveAtEnd(survivalOutcome)
}

final case class ReplayRecord(
  replayId: ReplayId,
  battleId: BattleId,
  handle: PlayerHandle,
  displayName: DisplayName,
  finishedAt: EpochMillis,
  finishedAtLabel: String,
  title: ReplayTitle,
  modeLabel: String,
  resultLabel: String,
  mapLabel: String,
  highlightLine: String,
  coverLabel: String,
  playersLine: String,
  timelineHint: String,
  score: Score,
  placement: Option[BattlePlacement],
  ratingBefore: Option[Rating],
  ratingDelta: Option[RatingDelta],
  ratingAfter: Option[Rating],
  durationMs: DurationMillis,
  survivalOutcome: BattleSurvivalOutcome,
  thumbnailDataUrl: Option[String],
  currentLoadout: Option[String],
  frameCount: ReplayFrameCount,
  playbackAvailability: ReplayPlaybackAvailability,
  framesJson: ReplayFramesJson,
  settlements: Vector[ReplaySettlementRecord] = Vector.empty
) {
  def aliveAtEnd: Boolean =
    BattleSurvivalOutcome.aliveAtEnd(survivalOutcome)

  def playbackAvailable: Boolean =
    ReplayPlaybackAvailability.availableFlag(playbackAvailability)

  def settlementFor(handle: PlayerHandle): Option[ReplaySettlementRecord] =
    settlements.find(_.handle.key == handle.key)
}

final case class ReplayCommentRecord(
  id: ReplayCommentId,
  replayId: ReplayId,
  authorHandle: PlayerHandle,
  body: String,
  createdAt: EpochMillis
)

File: frontend/src/objects/replay/replayTypes.ts
import type {
  BattleReplayFrameState,
  BattleReplayHeroFrameState,
  BattleReplayPickupFrameState,
  BattleReplayProjectileFrameState
} from "../battle/microservices/projections/objects/replay/BattleReplayFrameState";

export type ReplayHeroFrame = BattleReplayHeroFrameState;
export type ReplayProjectileFrame = BattleReplayProjectileFrameState;
export type ReplayPickupFrame = BattleReplayPickupFrameState;
export type ReplayFrame = BattleReplayFrameState;

export interface ReplayPlayback {
  id: string;
  playbackAvailable?: boolean;
  title: string;
  modeLabel: string;
  resultLabel: string;
  finishedAtLabel: string;
  mapLabel: string;
  highlightLine: string;
  timelineHint: string;
  playersLine: string;
  score: number;
  placement: number | null;
  ratingBefore: number | null;
  ratingAfter: number | null;
  ratingDelta: number | null;
  durationMs: number;
  aliveAtEnd: boolean;
  thumbnailDataUrl: string | null;
  frames: ReplayFrame[];
}

export interface ReplayExportArtifact {
  filename: string;
  json: string;
}

export interface ReplayCatalogItemResponseDto {
  replayId: string;
  battleId: string;
  title: string;
  modeLabel: string;
  resultLabel: string;
  finishedAt: number;
  finishedAtLabel: string;
  mapLabel: string;
  highlightLine: string;
  coverLabel: string;
  playersLine: string;
  timelineHint: string;
  score: number;
  placement: number | null;
  ratingBefore: number | null;
  ratingAfter: number | null;
  ratingDelta: number | null;
  durationMs: number;
  aliveAtEnd: boolean;
  thumbnailDataUrl: string | null;
  frameCount: number;
  playbackAvailable: boolean;
}

export interface ReplayCatalogResponseDto {
  replays: ReplayCatalogItemResponseDto[];
}

export interface ReplayDetailRecordResponseDto extends ReplayCatalogItemResponseDto {
  handle: string;
  displayName: string;
  currentLoadout: string | null;
  frames: ReplayFrame[];
}

export interface ReplayDetailResponseDto {
  replay: ReplayDetailRecordResponseDto;
}

export interface ReplayCommentResponseDto {
  id: string;
  replayId: string;
  authorHandle: string;
  body: string;
  createdAt: number;
}

export interface ReplayCommentsResponseDto {
  comments: ReplayCommentResponseDto[];
}

export interface ReplayCommentWrapperResponseDto {
  comment: ReplayCommentResponseDto;
}

export interface ReplayRecordApiRequestDto {
  replayId?: string;
  battleId?: string;
  handle?: string;
  displayName?: string;
  finishedAt?: number;
  finishedAtLabel?: string;
  title?: string;
  modeLabel?: string;
  resultLabel?: string;
  mapLabel?: string;
  highlightLine?: string;
  coverLabel?: string;
  playersLine?: string;
  timelineHint?: string;
  score?: number;
  placement?: number | null;
  durationMs?: number;
  aliveAtEnd?: boolean;
  thumbnailDataUrl?: string | null;
  currentLoadout?: string | null;
  frameCount?: number;
  playbackAvailable?: boolean;
  framesJson?: string;
  frames?: ReplayFrame[];
}

export interface ReplayCommentApiRequestDto {
  authorHandle?: string;
  body?: string;
}

export interface ReplayCatalogApiRequestDto {
  limit?: number;
  handle?: string;
}
